### PR TITLE
worker: set up iptables and cleanup in teardown

### PR DESCRIPTION
**Pull request by rcooke-warwick**
This puts the "required" iptables rules set up into the worker set up code, along with the appropriate teardowns. It also removes the bridge interface we create as part of the teardown.

These changes are needed to allow us to run the worker in jenkins instances cleanly

Change-type: patch
Signed-off-by: Ryan Cooke <ryan@balena.io>

### CODE REVIEW

File: worker/lib/workers/qemu.ts [hunk 1]
```diff
@@ -76,6 +80,19 @@ class QemuWorker extends EventEmitter implements Leviathan.Worker {
 		if (this.screenCapturer != null) {
 			await this.screenCapturer.teardown();
 		}
+
+		// remove iptables rules set up from host
+		try{
+			await execProm(`iptables-legacy-save | grep -v 'comment ${this.iptablesComment}' | iptables-legacy-restore`)
```
Comment (klutchell) [inline worker/lib/workers/qemu.ts:86]: Need to test this command (without the grep) on real jenkins-agent shell.

https://dashboard.balena-cloud.com/devices/11a1d0258e8782b188c3598b57c155c6/summary

File: worker/lib/workers/qemu.ts [hunk 2]
```diff
@@ -188,6 +205,12 @@ class QemuWorker extends EventEmitter implements Leviathan.Worker {
 		});
 	}
 
+	private async iptablesRules(bridgeName: string, bridgeAddress: string){
+		await execProm(`iptables-legacy -I FORWARD -i ${bridgeName} -j ACCEPT -m comment --comment ${this.iptablesComment}`);
+		await execProm(`iptables-legacy -I FORWARD -o ${bridgeName} -m state --state RELATED,ESTABLISHED -j ACCEPT -m comment --comment ${this.iptablesComment}`);
+		await execProm(`iptables-legacy -t nat -A POSTROUTING ! -o ${bridgeName} --source ${bridgeAddress}/24 -j MASQUERADE -m comment --comment ${this.iptablesComment}`);
```
Comment (klutchell) [inline worker/lib/workers/qemu.ts:211]: Need to test this command on real jenkins-agent shell in balena-jenkins fleet.

https://dashboard.balena-cloud.com/fleets/1841113/summary

File: worker/lib/workers/qemu.ts (hunk 1 above)
Comment (klutchell) [reply worker/lib/workers/qemu.ts:86]:
```
jenkins@f8d197dc699e:~$ iptables-legacy-save | iptables-legacy-restore
Failed to list table names in /proc/net/ip_tables_names: Permission denied
```

Comment (klutchell) [reply worker/lib/workers/qemu.ts:86]: Seems to work as root, just not as the jenkins user. This would have been a similar issue on your workstation so we must be running as root somewhere?
```
sudo iptables-legacy-save | sudo iptables-legacy-restore
```

Comment (klutchell) [reply worker/lib/workers/qemu.ts:86]: Can we also try to run this command on teardown? Ideally we would clean up rules on start and stop of the worker.

Comment (rcooke-warwick) [reply worker/lib/workers/qemu.ts:86]: @klutchell It works from within the worker container locally because we have this mapping in the docker-compose I believe:

```
"/dev/net/tun:/dev/net/tun"
```

But if the jenkins agent shell that is managing the container doesn't have the right permissions, then maybe that will still stop it working, I'll have a look

Comment (klutchell) [reply worker/lib/workers/qemu.ts:86]: If we are running `docker-compose up` from the jenkins agent and have permissions to do that, then I guess it doesn't matter what other permissions are available since it's now up to our compose file?

It looks like we are using docker-in-docker on the agent so we really just need to test it I think.
https://github.com/product-os/balena-jenkins/blob/master/docker-compose.yml#L192

Comment (jakogut) [reply worker/lib/workers/qemu.ts:86]: > It works from within the worker container locally because we have this mapping in the docker-compose I believe

The tun mapping is for QEMU TUN/TAP networking, it works locally because the container is using host mode networking with the NET_ADMIN capability

File: worker/lib/workers/qemu.ts [hunk 3]
```diff
@@ -76,6 +80,19 @@ class QemuWorker extends EventEmitter implements Leviathan.Worker {
...
+
+		// remove iptables rules set up from host
+		try{
+			await execProm(`iptables-legacy-save | grep -v 'comment ${this.iptablesComment}' | iptables-legacy-restore`)
+		} catch (e){
+			console.log(`error while removing iptables rules: ${e}`)
+		}
+
+		try{
+			await execProm(`ip link set dev ${this.qemuOptions.network.bridgeName} down && brctl delbr ${this.qemuOptions.network.bridgeName}`)
```
Comment (jakogut) [inline worker/lib/workers/qemu.ts:92]: Could we split this line up? It's quite long.

File: worker/lib/workers/qemu.ts (hunk 2 above)
Comment (jakogut) [inline worker/lib/workers/qemu.ts:211]: These rules could probably be simplified. The first two should be unnecessary, and the last rule could be simplified to `iptables-legacy -t nat -A POSTROUTING -j MASQUERADE -m comment ...`

We can optionally specify the output device, which should be the external interface, but this is not necessary.

File: worker/lib/workers/qemu.ts [hunk 4]
```diff
@@ -241,19 +264,19 @@ class QemuWorker extends EventEmitter implements Leviathan.Worker {
...
+		const bridgeName: string = this.qemuOptions.network.bridgeName;
+		const bridgeAddress: string = this.qemuOptions.network.bridgeAddress;
+		const dnsmasqArgs = [
+		`--interface=${bridgeName}`,
+		`--dhcp-range=${this.qemuOptions.network.dhcpRange}`,
+		'--conf-file',
+		'--bind-interfaces',
+		'--no-daemon',
+		`--dhcp-leasefile=/var/run/qemu-dnsmasq-${bridgeName}.leases`,
+		];
```
Comment (jakogut) [inline worker/lib/workers/qemu.ts:276]: Same as the above comment on diff noise, this should be in a separate commit, which makes things easier to review. :)

**Review by jakogut — CHANGES_REQUESTED** [review]
Thanks for working on this. I'm wondering if we should also at least check that forwarding is enabled, and maybe print an error message if it's not.

**Comment by klutchell** [comment]
@balena-ci rebase

**Comment by klutchell** [comment]
Rebased so we have the latest Makefile changes from master (`make detached`).

File: worker/lib/workers/qemu.ts (hunk 2 above)
Comment (rcooke-warwick) [reply worker/lib/workers/qemu.ts]: made the change and tested this locally - works well thanks :)

Comment (rcooke-warwick) [reply worker/lib/workers/qemu.ts]: works within jenkins jobs because the container is using host mode networking with the NET_ADMIN capability

Comment (rcooke-warwick) [reply worker/lib/workers/qemu.ts]: @jakogut , I removed the first 2 rules, but for the last rule I had to specify the interface, otherwise my laptop couldn't connect to the internet while the rule was active.

**Comment by klutchell** [comment]
@balena-ci rebase

File: worker/lib/workers/qemu.ts [hunk 5]
```diff
@@ -130,16 +130,9 @@ class QemuWorker extends EventEmitter implements Leviathan.Worker {
...
-				['resize', '-f', 'raw', this.image, '8G']
-			);
-
-			qemuImgProc.on(
-				'error', (err) => {
-					reject(err);
-				}
-			);
+			console.log(`Resizing qemu image...`)
+			await execProm(`qemu-img resize -f raw ${this.image} 8G`)
```
Comment (klutchell) [inline worker/lib/workers/qemu.ts:134]: Does execProm catch errors?

File: worker/lib/workers/qemu.ts [hunk 6]
```diff
@@ -76,6 +80,30 @@ class QemuWorker extends EventEmitter implements Leviathan.Worker {
...
+			await execProm(`ip link set dev ${this.qemuOptions.network.bridgeName} down`);
+			await execProm(`brctl delbr ${this.qemuOptions.network.bridgeName}`);
+		} catch(e){
+			console.log(`error while removing bridge: ${e}`)
+		}
+
+		await new Promise((resolve, reject) => {
+			if (this.dnsmasqProc && !this.dnsmasqProc.killed) {
+				// don't return until the process is dead
+				!this.dnsmasqProc.on('exit', resolve)
```
Comment (jakogut) [inline worker/lib/workers/qemu.ts:101]: What is the `!` here for? Also, missing semicolon.

File: worker/lib/workers/qemu.ts (hunk 5 above)
Comment (rcooke-warwick) [reply worker/lib/workers/qemu.ts:134]: It does @klutchell yeah

File: worker/lib/workers/qemu.ts (hunk 6 above)
Comment (rcooke-warwick) [reply worker/lib/workers/qemu.ts]: Thats a typo, well spotted

**Review by klutchell — APPROVED** [review]
LGTM, we just need @jakogut to sign off on the requested changes

**Review by jakogut — COMMENTED** [review]
On line 70, the sysctl command should have `=1` at the end. There's an unresolved comment on line 109 about resolve not being called (on the line above, it's being passed as a callback).

**Review by jakogut — APPROVED** [review]
Looks good, thanks for addressing feedback!
